### PR TITLE
[SCHEMA] Add common principles to schema

### PR DESCRIPTION
**Pull request by tsalo**
Closes None.

Changes proposed:
- Add `schema/objects/common_principles.yaml` with descriptions of common principles.
- Add `schema/rules/common_principles.yaml` with the order of common principles as they should appear in the specification.
- Add a rendering function and associated macro to print the common principles in the specification.

### CODE REVIEW

**Comment by tsalo** [comment]
The only thing I'm having trouble with is keeping the text after the nested list in the data type definition within the overall common principles list. At the moment, it is rendered outside the list and the list is restarted with the next item (task).
image

**Comment by Remi-Gau** [comment]
had missed this one.

let me have a quick look

File: src/schema/objects/common_principles.yaml [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    A set of neuroimaging and behavioral data acquired for a purpose of a particular study.
+    A dataset consists of data acquired from one or more subjects, possibly from multiple sessions.
+deprecated:
+  name: DEPRECATED
+  display_name: DEPRECATED
+  description: |
+    A "deprecated" entity or metadata field SHOULD NOT be used in the generation of new datasets.
+    It remains in the standard in order to preserve the interpretability of existing datasets.
+    Validating software SHOULD warn when deprecated practices are detected and provide a suggestion for updating the
+    dataset to preserve the curator's intent.
```
Comment (Remi-Gau) [inline src/schema/objects/common_principles.yaml:55]:
```suggestion
    Validating software SHOULD warn when deprecated practices are detected 
    and provide a suggestion for updating the dataset to preserve the curator's intent.
```

File: src/schema/objects/common_principles.yaml [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+event:
+  name: Event
+  display_name: Event
+  description: |
+    Something that happens or may be perceived by a test subject as happening at a particular instant during the
+    recording.
```
Comment (Remi-Gau) [inline src/schema/objects/common_principles.yaml:61]:
```suggestion
    Something that happens or may be perceived by a test subject as happening 
    at a particular instant during the recording.
```

File: src/schema/objects/common_principles.yaml [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+    Events are most commonly associated with on- or offset of stimulus presentations,
+    or with the distinct marker of on- or offset of a subject's response or motor action.
+    Other events may include unplanned incidents
+    (for example, sudden onset of noise and vibrations due to construction work, laboratory device malfunction),
+    changes in task instructions (for example, switching the response hand),
```
Comment (Remi-Gau) [inline src/schema/objects/common_principles.yaml:66]:
```suggestion
    (for example, sudden onset of noise and vibrations due to construction work, 
    laboratory device malfunction),
    changes in task instructions (for example, switching the response hand),
```

File: src/schema/objects/common_principles.yaml [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    and noted data feature occurrences (for example, a recording electrode producing noise).
+    In BIDS, each event has an onset time and duration.
+    Note that not all tasks will have recorded events (for example, "resting state").
+extension:
+  name: File extension
+  display_name: File extension
+  description: |
+    A portion of the file name after the left-most period (`.`) preceded by any other alphanumeric.
+    For example, `.gitignore` does not have a file extension, but the file extension of `test.nii.gz` is `.nii.gz`.
+    Note that the left-most period is included in the file extension.
```
Comment (Remi-Gau) [inline src/schema/objects/common_principles.yaml:77]:
```suggestion
    For example, `.gitignore` does not have a file extension, 
    but the file extension of `test.nii.gz` is `.nii.gz`.
    Note that the left-most period is included in the file extension.
```

File: src/schema/objects/common_principles.yaml [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+    for example, it is `rest` in `task-rest` following `task-<label>` specification.
+    Note that labels MUST not collide when casing is ignored
+    (see [Case collision intolerance](SPEC_ROOT/02-common-principles.md#case-collision-intolerance)).
+modality:
+  name: Modality
+  display_name: Modality
+  description: |
+    The category of brain data recorded by a file.
+    For MRI data, different pulse sequences are considered distinct modalities, such as `T1w`, `bold` or `dwi`.
+    For passive recording techniques, such as EEG, MEG or iEEG,
```
Comment (Remi-Gau) [inline src/schema/objects/common_principles.yaml:98]:
```suggestion
    For MRI data, different pulse sequences are considered distinct modalities, 
    such as `T1w`, `bold` or `dwi`.
    For passive recording techniques, such as EEG, MEG or iEEG,
```

File: src/schema/objects/common_principles.yaml [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+    the technique is sufficiently uniform to define the modalities `eeg`, `meg` and `ieeg`.
+    When applicable, the modality is indicated in the **suffix**.
+    The modality may overlap with, but should not be confused with the **data type**.
+run:
+  name: Run
+  display_name: Run
+  description: |
+    An uninterrupted repetition of data acquisition that has the same acquisition parameters and task
+    (however events can change from run to run due to different subject response or randomized nature of the stimuli).
+    Run is a synonym of a data acquisition.
```
Comment (Remi-Gau) [inline src/schema/objects/common_principles.yaml:108]:
```suggestion
    (however events can change from run to run due to different subject response 
    or randomized nature of the stimuli).
    Run is a synonym of a data acquisition.
```

File: src/schema/objects/common_principles.yaml [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+    if a subject leaves the scanner, the acquisition must be restarted.
+    For some types of [PET](SPEC_ROOT/04-modality-specific-files/09-positron-emission-tomography.md) acquisitions,
+    a subject may leave and re-enter the scanner without interrupting the scan.
+sample:
+  name: Sample
+  display_name: Sample
+  description: |
+    A sample pertaining to a subject such as tissue, primary cell or cell-free sample.
+    Sample labels MUST be unique within a subject and it is RECOMMENDED that they be unique throughout the dataset.
+session:
```
Comment (Remi-Gau) [inline src/schema/objects/common_principles.yaml:121]:
```suggestion
    Sample labels MUST be unique within a subject and it is RECOMMENDED
    that they be unique throughout the dataset.
session:
```

File: src/schema/objects/common_principles.yaml [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+  name: Session
+  display_name: Session
+  description: |
+    A logical grouping of neuroimaging and behavioral data consistent across subjects.
+    Session can (but doesn't have to) be synonymous to a visit in a longitudinal study.
+    In general, subjects will stay in the scanner during one session.
+    However, for example, if a subject has to leave the scanner room and then be re-positioned on the scanner bed,
+    the set of MRI acquisitions will still be considered as a session and match sessions acquired in other subjects.
```
Comment (Remi-Gau) [inline src/schema/objects/common_principles.yaml:129]:
```suggestion
    However, for example, if a subject has to leave the scanner room 
    and then be re-positioned on the scanner bed,
    the set of MRI acquisitions will still be considered as a session 
    and match sessions acquired in other subjects.
```

File: src/schema/objects/common_principles.yaml [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+    Similarly, in situations where different data types are obtained over several visits
+    (for example fMRI on one day followed by DWI the day after) those can be grouped in one session.
+    Defining multiple sessions is appropriate when several identical or similar data acquisitions are planned and
+    performed on all -or most- subjects,
```
Comment (Remi-Gau) [inline src/schema/objects/common_principles.yaml:133]:
```suggestion
    Defining multiple sessions is appropriate when several identical or similar data acquisitions 
    are planned and performed on all -or most- subjects,
```

File: src/schema/rules/common_principles.yaml [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+- dataset
+- subject
+- session
+- sample
+- data_acquisition
+- data_type
+- task
+- event
+- run
+- modality
```
Comment (Remi-Gau) [inline src/schema/rules/common_principles.yaml:13]:
```suggestion
- dataset
- modality
- data_type
- subject
- session
- sample
- data_acquisition
- task
- event
- run
```

Comment (Remi-Gau) [reply src/schema/rules/common_principles.yaml:13]: Trying to order this from "general" to "specific", I would bump modality and data_type higher up.

Not a hill I will die on.

**Review by Remi-Gau — CHANGES_REQUESTED** [review]
Minor formatting issues (attempt at semantic linebreaks here and there).

Also small suggestion to reorder definitions in common principles.

File: src/schema/rules/common_principles.yaml (hunk 10 above)
Comment (tsalo) [reply src/schema/rules/common_principles.yaml:13]: I was just replicating the existing organization, but I do agree that the principles could be reordered to be more informative. General to specific seems like the best way to go! 👍

**Comment by tsalo** [comment]
Thanks for your review @Remi-Gau!

**Review by sappelhoff — APPROVED** [review]
@Remi-Gau do you approve?

**Review by Remi-Gau — APPROVED** [review]
@sappelhoff : :+1: 
@tsalo : Thanks for doing this. :rocket:

**Pull request merged**